### PR TITLE
Oauth2

### LINKED ISSUE #46: Implement OAuth2.0 PKCE flow
**Is your feature request related to a problem? Please describe.**
No

**Describe the solution you'd like**
Implement OAuth2.0 PKCE flow based on the following RFCs

- https://datatracker.ietf.org/doc/html/rfc6749
- https://datatracker.ietf.org/doc/html/rfc7636

**Describe alternatives you've considered**
Session cookie, but not interesting enough and I have too much time

**Additional context**
N/A

**Todo List**
- [x] `/signup` endpoint for adding new user
- [x] `/oauth/authorize` endpoint for verifying the user
- [x] `/oauth/token` endpoint for getting JWT
- [x] `/oauth/refresh` endpoint for refreshing JWT
- [x] `IsRole()` middleware for verifying JWT
- [x] add `/login` frontend path for initiating PKCE flow
- [x] add `/authorize` frontend path for login input
- [x] add `/callback` frontend path for token exchange

### PR DESCRIPTION
**Pull request by NOOBDY**
Resolves #46 

### Changes

TODO Listed in issue

### Testing Approach (if unconventional)

- Create a user in the DB (form validation in `/signup` isn't implemented)
- Go to `/login` and click on the button
- Enter your login credentials
- After being redirected to the home page you can navigate to `/APITest` and you should see "pong" on the screen
- Clicking on "Logout" inside the drop down in the top right corner would redirect you to the login page
- Navigating to any path inside the `<ProtectedBoundary />` component would also redirect you to the login page

### Additional context

N/A

### CODE REVIEW

File: db/queries/user.sql [hunk 1]
```diff
@@ -71,3 +73,29 @@ SELECT EXISTS (
...
             OR "email" = $2
     );
+
+-- name: FindUserInfoAndPassword :one
+SELECT "username",
+    "role",
+    "password"
+FROM "user"
+WHERE "username" = $1
+    OR "email" = $1;
```
Comment (Error0229) [inline db/queries/user.sql:85]: since it seems only allow login with email, the sql only need email as input

**Comment by Error0229** [comment]
Test failed when clicked login button, got `ERR_CONNECTION_REFUSED`
Request
```
curl 'http://localhost:5173/authorize?client_id=twp&code_challenge=p0-SBpejdZ0AuRVuQxSsmw5u25j9Ra05JF8E5uFPFv0&code_challenge_method=S256&redirect_uri=http%3A%2F%2Flocalhost%3A8080%2Fcallback&response_type=code&state=NjQzNTgxNjU4MzU5NDMwNTQ4OTI1NjIzMjUxNzExNzIzNDg0NjUyNTA2NjMxNzA5MzIwNDg0ODQzMTk2NDY1MDYwODM2MDkyNTUxODU' \
  -H 'Referer: http://localhost:8080/' \
  -H 'Upgrade-Insecure-Requests: 1' \
  -H 'User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36 Edg/120.0.0.0' \
  -H 'sec-ch-ua: "Not_A Brand";v="8", "Chromium";v="120", "Microsoft Edge";v="120"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Windows"' \
  --compressed
```

Log
```
twp-twp-1       | {"level":"error","ts":1702316687.702742,"caller":"auth/refresh.go:17","msg":"http: named cookie not present","stacktrace":"github.com/jykuo-love-shiritori/twp/pkg/router.RegisterApi.Refresh.func7\n\t/app/pkg/auth/refresh.go:17\ngithub.com/labstack/echo/v4.(*Echo).add.func1\n\t/go/pkg/mod/github.com/labstack/echo/v4@v4.11.2/echo.go:582\ngithub.com/labstack/echo/v4/middleware.StaticWithConfig.func1.1\n\t/go/pkg/mod/github.com/labstack/echo/v4@v4.11.2/middleware/static.go:167\ngithub.com/labstack/echo/v4/middleware.RequestLoggerConfig.ToMiddleware.func1.1\n\t/go/pkg/mod/github.com/labstack/echo/v4@v4.11.2/middleware/request_logger.go:259\ngithub.com/labstack/echo/v4.(*Echo).ServeHTTP\n\t/go/pkg/mod/github.com/labstack/echo/v4@v4.11.2/echo.go:669\nnet/http.serverHandler.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2938\nnet/http.(*conn).serve\n\t/usr/local/go/src/net/http/server.go:2009"}
twp-twp-1       | {"level":"info","ts":1702316687.7028463,"caller":"app/logger.go:15","msg":"request","method":"POST","URI":"/api/oauth/refresh","status":401}
twp-twp-1       | {"level":"info","ts":1702316687.7036245,"caller":"app/logger.go:15","msg":"request","method":"GET","URI":"/assets/logo-28e599d4.png","status":200}
twp-twp-1       | {"level":"error","ts":1702316687.7264962,"caller":"auth/refresh.go:17","msg":"http: named cookie not present","stacktrace":"github.com/jykuo-love-shiritori/twp/pkg/router.RegisterApi.Refresh.func7\n\t/app/pkg/auth/refresh.go:17\ngithub.com/labstack/echo/v4.(*Echo).add.func1\n\t/go/pkg/mod/github.com/labstack/echo/v4@v4.11.2/echo.go:582\ngithub.com/labstack/echo/v4/middleware.StaticWithConfig.func1.1\n\t/go/pkg/mod/github.com/labstack/echo/v4@v4.11.2/middleware/static.go:167\ngithub.com/labstack/echo/v4/middleware.RequestLoggerConfig.ToMiddleware.func1.1\n\t/go/pkg/mod/github.com/labstack/echo/v4@v4.11.2/middleware/request_logger.go:259\ngithub.com/labstack/echo/v4.(*Echo).ServeHTTP\n\t/go/pkg/mod/github.com/labstack/echo/v4@v4.11.2/echo.go:669\nnet/http.serverHandler.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2938\nnet/http.(*conn).serve\n\t/usr/local/go/src/net/http/server.go:2009"}
twp-twp-1       | {"level":"info","ts":1702316687.7265587,"caller":"app/logger.go:15","msg":"request","method":"POST","URI":"/api/oauth/refresh","status":401}
twp-twp-1       | {"level":"info","ts":1702316687.729732,"caller":"app/logger.go:15","msg":"request","method":"GET","URI":"/assets/login-22840771.jpg","status":200}
```

File: db/queries/user.sql (hunk 1 above)
Comment (Error0229) [reply db/queries/user.sql:85]: consider add some comment

**Review by Error0229 — CHANGES_REQUESTED** [review]

File: pkg/auth/token.go [hunk 2]
```diff
@@ -31,46 +34,87 @@ func Token(db *db.DB, logger *zap.SugaredLogger) echo.HandlerFunc {
 			return echo.NewHTTPError(http.StatusBadRequest, "Failed to parse data")
 		}
 
-		mu.RLock()
-		challenge := codeChallengePairs[params.Code]
-		mu.RUnlock()
+		fmt.Printf("%+v\n", c.QueryParams())
+		fmt.Printf("%+v\n", params)
```
Comment (Error0229) [inline pkg/auth/token.go]: consider remove debug codes

File: pkg/auth/verify_test.go
Comment (Error0229) [inline pkg/auth/verify_test.go:1]: should this move to `test/auth`?

**Review by Error0229 — CHANGES_REQUESTED** [review]

File: pkg/auth/verify_test.go
Comment (NOOBDY) [reply pkg/auth/verify_test.go:1]: this is a unit test that can be ran without human intervention or in CI, the `./test/...` are integration tests meant to be ran manually and sometimes would require interactions

File: pkg/auth/token.go (hunk 2 above)
Comment (NOOBDY) [reply pkg/auth/token.go]: fixed

**Review by Error0229 — DISMISSED** [review]
Backend part LGTM

**Comment by Error0229** [comment]
Is it normal that the invalid email/password login redirect to `http://localhost:8080/callback?state=undefined&code=undefined` ?Also that path is a empty page
The valid email/password login functions are works fine anyway.

**Comment by NOOBDY** [comment]
the `/callback` page is supposed to be empty, but I don't know what the behavior should be when the user inputs so I'll leave that to the frontend team

**Comment by UGisBusy** [comment]
I think the frontend can provide least amount of verification, i.e. email most contain @, every slot must not be empty.
This way gets little more UX then what it is right now imo.

**Review by Error0229 — APPROVED** [review]

**Review by UGisBusy — APPROVED** [review]
lgtm

**Pull request merged**